File: examples/ruby/schnorr-zkp/SchnorrZKP.runar.rb
require 'runar'

class SchnorrZKP < Runar::SmartContract
  prop :pub_key, Point

  def initialize(pub_key)
    super(pub_key)
    @pub_key = pub_key
  end

  runar_public r_point: Point, s: Bigint, e: Bigint
  def verify(r_point, s, e)
    assert ec_on_curve(r_point)
    s_g = ec_mul_gen(s)
    e_p = ec_mul(@pub_key, e)
    rhs = ec_add(r_point, e_p)
    assert ec_point_x(s_g) == ec_point_x(rhs)
    assert ec_point_y(s_g) == ec_point_y(rhs)
  end
end

File: examples/zig/schnorr-zkp/SchnorrZKP.runar.zig
const runar = @import("runar");

pub const SchnorrZKP = struct {
    pub const Contract = runar.SmartContract;

    pubKey: runar.Point,

    pub fn init(pubKey: runar.Point) SchnorrZKP {
        return .{ .pubKey = pubKey };
    }

    pub fn verify(self: *const SchnorrZKP, rPoint: runar.Point, s: runar.Bigint) void {
        runar.assert(runar.ecOnCurve(rPoint));

        const e = runar.bin2num(runar.hash256(runar.cat(rPoint, self.pubKey)));
        const sG = runar.ecMulGen(s);
        const eP = runar.ecMul(self.pubKey, e);
        const rhs = runar.ecAdd(rPoint, eP);

        runar.assert(runar.bytesEq(runar.ecEncodeCompressed(sG), runar.ecEncodeCompressed(rhs)));
    }
};
